## Importing necessary libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

##### Importing dataset

In [2]:
df = pd.read_csv(r"C:\Users\hp\Desktop\household_power_project\household_power_consumption.csv")

In [3]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
0,16/12/2006,17:24:00,4.216,0.418,234.84,18.4,0,1,17.0
1,16/12/2006,17:25:00,5.36,0.436,233.63,23,0,1,16.0
2,16/12/2006,17:26:00,5.374,0.498,233.29,23,0,2,17.0
3,16/12/2006,17:27:00,5.388,0.502,233.74,23,0,1,17.0
4,16/12/2006,17:28:00,3.666,0.528,235.68,15.8,0,1,17.0


##### Check shape

In [4]:
df.shape

(1048575, 9)

##### Check information

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 9 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   Date                   1048575 non-null  str    
 1   Time                   1048575 non-null  str    
 2   Global_active_power    1048575 non-null  str    
 3   Global_reactive_power  1048575 non-null  str    
 4   Voltage                1048575 non-null  str    
 5   Global_intensity       1048575 non-null  str    
 6   Sub_metering_1         1048575 non-null  str    
 7   Sub_metering_2         1048575 non-null  str    
 8   Sub_metering_3         1044506 non-null  float64
dtypes: float64(1), str(8)
memory usage: 107.7 MB


##### Statistical analysis

###### for numerical data

In [8]:
df.describe()

,Sub_metering_3
count,1.044506e+06
mean,5.933837e+00
std,8.210398e+00
min,0.000000e+00
25%,0.000000e+00
50%,0.000000e+00
75%,1.700000e+01
max,3.100000e+01


###### for categorical data

In [9]:
df.describe(include='object')

C:\Users\hp\AppData\Local\Temp\ipykernel_14780\87514550.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='object')


,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2
count,1048575,1048575,1048575,1048575,1048575,1048575,1048575,1048575
unique,729,1440,4076,492,2605,219,82,80
top,17/12/2006,17:24:00,0.218,0,?,1,0,0
freq,1440,729,6714,266300,4069,95350,958708,746737


###### Checking the data types

In [10]:
df.dtypes

Date                         str
Time                         str
Global_active_power          str
Global_reactive_power        str
Voltage                      str
Global_intensity             str
Sub_metering_1               str
Sub_metering_2               str
Sub_metering_3           float64
dtype: object

###### Data type conversion

###### Converting the type of Global_active_power,Global_reactive_power,Voltage into float

In [5]:
df['Global_active_power'].astype('float')

ValueError: could not convert string to float: '?'

###### Checking for the rows having "?"

In [6]:
df[df['Global_active_power']=="?"]

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
6839,21/12/2006,11:23:00,?,?,?,?,?,?,NaN
6840,21/12/2006,11:24:00,?,?,?,?,?,?,NaN
19724,30/12/2006,10:08:00,?,?,?,?,?,?,NaN
19725,30/12/2006,10:09:00,?,?,?,?,?,?,NaN
41832,14/1/2007,18:36:00,?,?,?,?,?,?,NaN
...,...,...,...,...,...,...,...,...,...
1043669,10/12/2008,11:53:00,?,?,?,?,?,?,NaN
1043670,10/12/2008,11:54:00,?,?,?,?,?,?,NaN
1043671,10/12/2008,11:55:00,?,?,?,?,?,?,NaN
1043672,10/12/2008,11:56:00,?,?,?,?,?,?,NaN


In [7]:
df[df['Global_active_power']=="?"].shape

(4069, 9)

1. As, we can see there are 4069 rows which are having complete "?"
2. So,we are dropping

In [8]:
df = df[df['Global_active_power']!="?"]

In [9]:
df[df['Global_active_power']=="?"]

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3


###### Checking the data type again

df.dtypes

In [15]:
df.columns

Index(['Date', 'Time', 'Global_active_power', 'Global_reactive_power',
       'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2',
       'Sub_metering_3'],
      dtype='str')

In [10]:
Num_data = ['Global_active_power', 'Global_reactive_power',
       'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2',
       'Sub_metering_3']

In [11]:
for i in Num_data:
    df[i] = df[i].astype('float')

In [18]:
df.dtypes

Date                         str
Time                         str
Global_active_power      float64
Global_reactive_power    float64
Voltage                  float64
Global_intensity         float64
Sub_metering_1           float64
Sub_metering_2           float64
Sub_metering_3           float64
dtype: object

###### Converting the date column into Datetime 

In [12]:
df['Date'] =pd.to_datetime(df['Date'])

C:\Users\hp\AppData\Local\Temp\ipykernel_13516\3437490821.py:1: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df['Date'] =pd.to_datetime(df['Date'])


In [20]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
0,2006-12-16,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0
1,2006-12-16,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0
2,2006-12-16,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0
3,2006-12-16,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0
4,2006-12-16,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0


In [21]:
df.dtypes

Date                     datetime64[us]
Time                                str
Global_active_power             float64
Global_reactive_power           float64
Voltage                         float64
Global_intensity                float64
Sub_metering_1                  float64
Sub_metering_2                  float64
Sub_metering_3                  float64
dtype: object

In [22]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
0,2006-12-16,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0
1,2006-12-16,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0
2,2006-12-16,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0
3,2006-12-16,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0
4,2006-12-16,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0


In [23]:
df['Sub_metering_1'].unique()

array([ 0.,  1.,  2.,  6., 37., 36., 38., 25., 12., 11.,  9.,  4., 33.,
        5., 34.,  8., 26., 27., 15., 20., 21., 23., 16., 10., 39., 32.,
       19., 35., 30., 40.,  7., 31., 17.,  3., 29., 24., 14., 41., 53.,
       76., 75., 77., 74., 28., 68., 58., 62., 13., 18., 22., 64., 71.,
       56., 72., 45., 42., 59., 66., 43., 44., 48., 46., 73., 55., 60.,
       70., 63., 67., 47., 65., 51., 50., 69., 78., 57., 49., 61., 52.,
       54., 79., 80.])

In [24]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
0,2006-12-16,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0
1,2006-12-16,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0
2,2006-12-16,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0
3,2006-12-16,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0
4,2006-12-16,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0


###### Extracting the Day_Name,Month_Name, and year

In [13]:
df['Day_Name']= df['Date'].dt.day_name()
df['Month_Name']= df['Date'].dt.month_name()
df['Year']= df['Date'].dt.year

In [26]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3,Day_Name,Month_Name,Year
0,2006-12-16,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0,Saturday,December,2006
1,2006-12-16,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0,Saturday,December,2006
2,2006-12-16,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0,Saturday,December,2006
3,2006-12-16,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0,Saturday,December,2006
4,2006-12-16,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0,Saturday,December,2006


##### Splitting the hrs and mins from time column

In [14]:
df[['Hours','Minutes','Seconds']]=df['Time'].str.split(':',expand=True)

In [15]:
df.drop(columns = ['Seconds'],inplace=True)

In [29]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3,Day_Name,Month_Name,Year,Hours,Minutes
0,2006-12-16,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0,Saturday,December,2006,17,24
1,2006-12-16,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0,Saturday,December,2006,17,25
2,2006-12-16,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0,Saturday,December,2006,17,26
3,2006-12-16,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0,Saturday,December,2006,17,27
4,2006-12-16,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0,Saturday,December,2006,17,28


###### Check for the duplicate rows

In [16]:
df.duplicated().sum()

np.int64(0)

###### Check for the null values

In [31]:
df.isnull().sum()

Date                     0
Time                     0
Global_active_power      0
Global_reactive_power    0
Voltage                  0
Global_intensity         0
Sub_metering_1           0
Sub_metering_2           0
Sub_metering_3           0
Day_Name                 0
Month_Name               0
Year                     0
Hours                    0
Minutes                  0
dtype: int64

In [32]:
df.columns

Index(['Date', 'Time', 'Global_active_power', 'Global_reactive_power',
       'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2',
       'Sub_metering_3', 'Day_Name', 'Month_Name', 'Year', 'Hours', 'Minutes'],
      dtype='str')

##### Create new column Total_metering for total consumption

In [17]:
df['Total_metering'] = df[['Sub_metering_1','Sub_metering_2','Sub_metering_3']].sum(axis=1)

df.head()

In [18]:
df.to_csv('Household_clean_data.csv',index=False)